File: app/common/rpc/codec.py
from models.game import MoveResult
from rpc import gameapi_pb2
from models.enums import PlayerRegistration, PlayerStatus, MoveStatus

def decodeRegisterPlayerResponse(response: gameapi_pb2.RegisterPlayerResponse):
    match response.registrationStatus:
        case gameapi_pb2.PlayerRegistration.REGISTRATION_SUCCESS:
            return PlayerRegistration.Success
        case gameapi_pb2.PlayerRegistration.NO_PLAYER_SLOT_LEFT:
            return PlayerRegistration.NoPlayerSlotsLeft
        case gameapi_pb2.PlayerRegistration.ALREADY_REGISTERED:
            return PlayerRegistration.AlreadyRegistered

def encodeRegisterPlayerResponse(registrationResult: PlayerRegistration):
    match registrationResult:
        case PlayerRegistration.Success:
            status = gameapi_pb2.PlayerRegistration.REGISTRATION_SUCCESS
        case PlayerRegistration.NoPlayerSlotsLeft:
            status = gameapi_pb2.PlayerRegistration.NO_PLAYER_SLOT_LEFT
        case PlayerRegistration.AlreadyRegistered:
            status = gameapi_pb2.PlayerRegistration.ALREADY_REGISTERED
    return gameapi_pb2.RegisterPlayerResponse(registrationStatus = status)

def decodeGetPossibleMovesResponse(response: gameapi_pb2.GetPossibleMovesResponse):
    result = []
    for move in response.moves:
        result.append((move.x, move.y))
    return result

def encodeGetPossibleMovesResponse(response):
    result = []
    for move in response:
        (x, y) = move
        result.append(gameapi_pb2.MoveCoordinates(x = x, y = y))
    return gameapi_pb2.GetPossibleMovesResponse(moves = result)

def decodeGetCurrentBoardResponse(response: gameapi_pb2.GetCurrentBoardResponse):
    result = []
    for row in response.grid:
        rows = []
        for item in row:
            rows.append(item)
        result.append(rows)
    return result

def encodeGetCurrentBoardResponse(response):
    result = []
    for row in response:
        result.append(gameapi_pb2.BoardRow(items = row.items))
    return gameapi_pb2.GetCurrentBoardResponse(grid = result)

def decodeCanMoveResponse(response: gameapi_pb2.CanMoveResponse):
    match response.playerStatus:
        case gameapi_pb2.PlayerStatus.WON:
            return PlayerStatus.Won
        case gameapi_pb2.PlayerStatus.LOST:
            return PlayerStatus.Lost
        case gameapi_pb2.PlayerStatus.DRAW:
            return PlayerStatus.Draw
        case gameapi_pb2.PlayerStatus.CAN_MOVE:
            return PlayerStatus.CanMove
        case gameapi_pb2.PlayerStatus.WAIT:
            return PlayerStatus.Wait
        case gameapi_pb2.PlayerStatus.UNREGISTERED_PLAYER:
            return PlayerStatus.UnregisteredPlayer

def encodeCanMoveResponse(response: PlayerStatus):
    match response:
        case PlayerStatus.Won:
            status = gameapi_pb2.PlayerStatus.WON
        case PlayerStatus.Lost:
            status = gameapi_pb2.PlayerStatus.LOST
        case PlayerStatus.Draw:
            status = gameapi_pb2.PlayerStatus.DRAW
        case PlayerStatus.CanMove:
            status = gameapi_pb2.PlayerStatus.CAN_MOVE
        case PlayerStatus.Wait:
            status = gameapi_pb2.PlayerStatus.WAIT
        case PlayerStatus.UnregisteredPlayer:
            status = gameapi_pb2.PlayerStatus.UNREGISTERED_PLAYER
    return gameapi_pb2.CanMoveResponse(playerStatus = status)

def encodePlayerNameRequest(playerName):
    return gameapi_pb2.PlayerNameRequest(playerName=playerName)

def encodeMoveRequest(playerName: str, index: int):
    return gameapi_pb2.MoveRequest(
        playerName=playerName, 
        index=index
    )

def decodeMoveResponse(response: gameapi_pb2.MoveResponse):
    match response.status:
        case gameapi_pb2.MoveStatus.MOVE_SUCCESS:
            status = MoveStatus.Success
        case gameapi_pb2.MoveStatus.INCORRECT:
            status = MoveStatus.Incorrect
        case _:
            status = MoveStatus.Error
    return MoveResult(status, response.move)

def encodeMoveResponse(response: MoveResult):
    match response.status:
        case MoveStatus.Success:
            status = gameapi_pb2.MoveStatus.MOVE_SUCCESS
        case MoveStatus.Incorrect:
            status = gameapi_pb2.MoveStatus.INCORRECT
        case _:
            status = gameapi_pb2.MoveStatus.ERROR
    return gameapi_pb2.MoveResponse(status = status, move = response.move)

File: ai_client/game_api/codec.go
package game_api

import (
	. "ai_client/common"
)

func ConvertPlayerRegistrationResponse(resp *RegisterPlayerResponse) RegistrationStatus {
	switch resp.RegistrationStatus {
	case PlayerRegistration_REGISTRATION_SUCCESS:
		return RegistrationSuccess
	case PlayerRegistration_NO_PLAYER_SLOT_LEFT:
		return NoPlayerSlotLeft
	case PlayerRegistration_ALREADY_REGISTERED:
		return AlreadyRegistered
	default:
		return RegistrationError
	}
}

func ConvertGetPossibleMovesResponse(resp *GetPossibleMovesResponse) []Coordinate {
	moves := resp.Moves
	result := make([]Coordinate, len(moves))
	for i, cord := range moves {
		result[i] = Coordinate{
			X: cord.X,
			Y: cord.Y,
		}
	}
	return result
}

func ConvertGetCurrentBoardResponse(resp *GetCurrentBoardResponse) [][]int32 {
	grid := resp.Grid
	result := make([][]int32, len(grid))
	for i, row := range grid {
		items := row.Items
		result[i] = make([]int32, len(items))
		for j, cell := range items {
			result[i][j] = cell
		}
	}
	return result
}

func ConvertCanMoveResponse(resp *CanMoveResponse) GameStatus {
	switch resp.PlayerStatus {
	case PlayerStatus_WON:
		return Won
	case PlayerStatus_LOST:
		return Lost
	case PlayerStatus_DRAW:
		return Draw
	case PlayerStatus_CAN_MOVE:
		return CanMove
	case PlayerStatus_WAIT:
		return Wait
	case PlayerStatus_UNREGISTERED_PLAYER:
		return UnregisteredPlayer
	default:
		return PlayerError
	}
}

func ConvertMoveResponse(resp *MoveResponse) GameMoveStatus {
	switch resp.Status {
	case MoveStatus_MOVE_SUCCESS:
		return MoveSuccess
	case MoveStatus_INCORRECT:
		return Incorrect
	default:
		return MoveError
	}
}
